sdk/identity/azure-identity/src/main/java/com/azure/identity/EnvironmentCredential.java

```diff
@@ -42,14 +42,30 @@
     @Override
     public Mono<AccessToken> getToken(TokenRequestContext request) {
         return Mono.fromSupplier(() -> {
-            if (configuration.contains(Configuration.PROPERTY_AZURE_CLIENT_ID)
-                && configuration.contains(Configuration.PROPERTY_AZURE_CLIENT_SECRET)
-                && configuration.contains(Configuration.PROPERTY_AZURE_TENANT_ID)) {
+            if (configuration.containsAll(Configuration.PROPERTY_AZURE_TENANT_ID,
+                    Configuration.PROPERTY_AZURE_CLIENT_ID,
+                    Configuration.PROPERTY_AZURE_CLIENT_SECRET)) {
                 // TODO: support other clouds
                 return new ClientSecretCredential(configuration.get(Configuration.PROPERTY_AZURE_TENANT_ID),
                     configuration.get(Configuration.PROPERTY_AZURE_CLIENT_ID),
                     configuration.get(Configuration.PROPERTY_AZURE_CLIENT_SECRET),
                     identityClientOptions);
+            } else if (configuration.containsAll(Configuration.PROPERTY_AZURE_CLIENT_ID,
+                    Configuration.PROPERTY_AZURE_TENANT_ID,
+                    Configuration.PROPERTY_AZURE_CLIENT_CERTIFICATE_PATH)) {
+                return new ClientCertificateCredential(configuration.get(Configuration.PROPERTY_AZURE_TENANT_ID),
+                    configuration.get(Configuration.PROPERTY_AZURE_CLIENT_ID),
+                    configuration.get(Configuration.PROPERTY_AZURE_CLIENT_CERTIFICATE_PATH),
+                    null,
+                    identityClientOptions);
+            } else if (configuration.containsAll(Configuration.PROPERTY_AZURE_USERNAME,
+                    Configuration.PROPERTY_AZURE_CLIENT_ID,
+                    Configuration.PROPERTY_AZURE_PASSWORD)) {
+                return new UsernamePasswordCredential(configuration.get(Configuration.PROPERTY_AZURE_CLIENT_ID),
+                    configuration.get(Configuration.PROPERTY_AZURE_TENANT_ID),
```

Comment:
Is this not explicitly needed by `UsernamePasswordCredential`? Otherwise the validation check is missing this.

Reply:
No it's not, but supply it anyway if the environment variable is set.